src/cudadev/CUDACore/GenericCachingAllocator.h

```diff
@@ -163,41 +172,36 @@ class GenericCachingAllocator {
       }
       searchKey = *iBlock;
       liveBlocks_.erase(iBlock);
+      device = Traits::memoryDevice(searchKey.device);
       cachedBytes_[device].live -= searchKey.bytes;
       cachedBytes_[device].liveRequested -= searchKey.bytesRequested;
 
       recache = (cachedBytes_[device].free + searchKey.bytes <= maxCachedBytes_);
       if (recache) {
+        Traits::recordEvent(searchKey.readyEvent, searchKey.associatedQueue);
         cachedBlocks_.insert(searchKey);
         cachedBytes_[device].free += searchKey.bytes;
 
         if (debug_) {
           std::cout << "\t" << Traits::printDevice(device) << " returned " << searchKey.bytes << " bytes at " << ptr
                     << " from associated queue " << searchKey.associatedQueue << " , event " << searchKey.readyEvent
-                    << " .\n\t\t " << cachedBlocks_.size()
-                    << " available "
-                       "blocks cached ("
-                    << cachedBytes_[device].free << " bytes), " << liveBlocks_.size() << " live blocks outstanding. ("
-                    << cachedBytes_[device].live << " bytes)" << std::endl;
+                    << " .\n\t\t " << cachedBlocks_.size() << " available blocks cached (" << cachedBytes_[device].free
+                    << " bytes), " << liveBlocks_.size() << " live blocks outstanding. (" << cachedBytes_[device].live
+                    << " bytes)" << std::endl;
         }
       }
-
-      if (recache) {
-        Traits::recordEvent(searchKey.readyEvent, searchKey.associatedQueue);
-      }
     }
 
     if (not recache) {
+      auto scopedSetDevice = Traits::setDevice(searchKey.device);
```

Comment:
I think this is not needed, as `cudaFree`/`cudaFreeHost` and `cudaEventDestroy` can be called even if the current device does not match the one associated to the memory block and event.

I've tested by forcing the call to set the wrong device, and everything works fine.